nexus/types/src/deployment/blueprint_diff.rs

```diff
@@ -11,26 +11,621 @@ use super::blueprint_display::{
     BpTable, BpTableColumn, BpTableData, BpTableRow, KvListWithHeading, KvPair,
 };
 use super::{
-    zone_sort_key, Blueprint, ClickhouseClusterConfig,
-    CockroachDbPreserveDowngrade, DiffBeforeClickhouseClusterConfig,
+    unwrap_or_none, zone_sort_key, BlueprintDatasetConfigDiff,
+    BlueprintDatasetsConfigDiff, BlueprintDiff, BlueprintMetadata,
+    BlueprintPhysicalDiskConfig, BlueprintPhysicalDisksConfigDiff,
+    BlueprintZoneConfigDiff, BlueprintZonesConfigDiff, ClickhouseClusterConfig,
+    CockroachDbPreserveDowngrade,
 };
-use diffus::Diffable;
+use daft::Diffable;
 use nexus_sled_agent_shared::inventory::ZoneKind;
 use omicron_common::api::external::Generation;
-use omicron_common::disk::DiskIdentity;
+use omicron_common::disk::{DatasetName, DiskIdentity};
 use omicron_uuid_kinds::OmicronZoneUuid;
 use omicron_uuid_kinds::SledUuid;
 use std::collections::{BTreeMap, BTreeSet};
 use std::fmt;
 
 use crate::deployment::blueprint_display::BpClickhouseKeepersTableSchema;
 use crate::deployment::{
-    BlueprintDatasetConfig, BlueprintDatasetsConfig, BlueprintMetadata,
+    Blueprint, BlueprintDatasetConfig, BlueprintDatasetsConfig,
     BlueprintPhysicalDisksConfig, BlueprintZoneConfig,
     BlueprintZoneDisposition, BlueprintZonesConfig,
-    CollectionDatasetIdentifier, DiffBeforeMetadata, ZoneSortKey,
+    CollectionDatasetIdentifier, ZoneSortKey,
 };
-use crate::external_api::views::SledState;
+
+// A wrapper type around a `daft` generated `BlueprintDiff that provides summary
+// data and direct access to the underlying diff.
+#[derive(Debug)]
+pub struct BlueprintDiffSummary<'a> {
+    pub before: &'a Blueprint,
+    pub after: &'a Blueprint,
+    pub diff: BlueprintDiff<'a>,
+    pub all_sleds: BTreeSet<SledUuid>,
+    pub sleds_added: BTreeSet<SledUuid>,
+    pub sleds_removed: BTreeSet<SledUuid>,
+    pub sleds_modified: BTreeSet<SledUuid>,
+    pub sleds_unchanged: BTreeSet<SledUuid>,
+}
+
+impl<'a> BlueprintDiffSummary<'a> {
+    pub fn new(before: &'a Blueprint, after: &'a Blueprint) -> Self {
+        let diff = before.diff(after);
+        // We assume for now that sled_state additions represent sled additions.
+        // Once we collapse the 4 blueprint maps this will be unambiguously
+        // true.
+        let sleds_added: BTreeSet<SledUuid> =
+            diff.sled_state.added.keys().map(|k| **k).collect();
+
+        // We can't do the same for removals unfortunately. We prematurely
+        // prune decommissioned sleds, but there may still be zones, disks,
+        // or datasets that have not yet been removed. We must check for this.
```

Comment:
I was going to ask if we could switch this to just looking at `sled_state` once #7470 lands, but I guess that would break diffs for blueprints involving decommissioned sleds made before that change landed. Maybe once we start pruning old blueprints, then?

Reply:
Yeah, all things that break diffs should come in follow ups.